ams/server/src/main/java/com/netease/arctic/server/table/TableSelfOptimizingMetrics.java

```diff
@@ -0,0 +1,350 @@
+      registerMetric(registry, TABLE_OPTIMIZING_MINOR_FAILED_COUNT, minorFailedCount);
+      registerMetric(registry, TABLE_OPTIMIZING_MAJOR_TOTAL_COUNT, majorTotalCount);
+      registerMetric(registry, TABLE_OPTIMIZING_MAJOR_FAILED_COUNT, majorFailedCount);
+      registerMetric(registry, TABLE_OPTIMIZING_FULL_TOTAL_COUNT, fullTotalCount);
+      registerMetric(registry, TABLE_OPTIMIZING_FULL_FAILED_COUNT, fullFailedCount);
+
+      globalRegistry = registry;
+    }
+  }
+
+  public void unregister() {
+    registeredMetricKeys.forEach(globalRegistry::unregister);
+    registeredMetricKeys.clear();
+    globalRegistry = null;
+  }
+
+  /**
+   * Handle table self optimizing state change event.
+   *
+   * @param optimizingStatus new optimizing status
+   * @param stateSetTimestamp timestamp of status changed.
+   */
+  public void stateChanged(OptimizingStatus optimizingStatus, long stateSetTimestamp) {
+    this.optimizingStatus = optimizingStatus;
+    this.stateSetTimestamp = stateSetTimestamp;
+  }
+
+  /**
+   * Handle table self optimizing process completed event.
+   *
+   * @param processType optimizing process type.
+   * @param success is optimizing process success.
+   */
+  public void processComplete(OptimizingType processType, boolean success) {
+    processTotalCount.inc();
+    Counter totalCounter = null;
+    Counter failedCounter = null;
+    switch (processType) {
+      case MINOR:
+        totalCounter = minorTotalCount;
+        failedCounter = minorFailedCount;
+        break;
+      case MAJOR:
+        totalCounter = majorTotalCount;
+        failedCounter = majorFailedCount;
+        break;
+      case FULL:
+        totalCounter = fullTotalCount;
+        failedCounter = fullFailedCount;
+        break;
+    }
+    if (totalCounter != null) {
+      totalCounter.inc();
+    }
+    if (!success && failedCounter != null) {
+      failedCounter.inc();
+    }
+  }
+
+  private String optimizingStatusToMetricState(OptimizingStatus status) {
+    switch (status) {
+      case IDLE:
+        return STATE_IDLE;
+      case PENDING:
+        return STATE_PENDING;
+      case PLANNING:
+        return STATE_PLANING;
+      case FULL_OPTIMIZING:
+      case MAJOR_OPTIMIZING:
+      case MINOR_OPTIMIZING:
+        return STATE_EXECUTING;
+      case COMMITTING:
+        return STATE_COMMITTING;
+      default:
+        return status.name();
+    }
+  }
+
+  class StateDurationGauge implements Gauge<Long> {
+    final String targetState;
+
+    StateDurationGauge(String targetState) {
+      this.targetState = targetState;
+    }
+
+    @Override
+    public Long getValue() {
+      String state = optimizingStatusToMetricState(optimizingStatus);
+      if (targetState.equals(state)) {
+        return stateDuration();
+      }
+      return 0L;
+    }
+
+    private Long stateDuration() {
+      return System.currentTimeMillis() - stateSetTimestamp;
+    }
+  }
+
+  class InStatusGauge implements Gauge<Long> {
```

Comment:
I can not get the difference between this and StateDurationGauge

Shall we make it more clear by name?